File: packages/honker-node/examples/atomic.js
// Atomic business-write + enqueue in one transaction.
//
// This is honker's killer feature vs Redis/BullMQ: the job enqueue
// commits in the same SQLite transaction as your business INSERT.
// Rollback drops both. No dual-write, no outbox table.
//
//   node examples/atomic.js

const lit = require('..');
const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');
const assert = require('node:assert/strict');

const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'honker-'));
const db = lit.open(path.join(dir, 'app.db'));

try {
  // One-time: bootstrap honker + create a plain business table.
  {
    const tx = db.transaction();
    tx.query("SELECT honker_bootstrap()");
    tx.execute(
      "CREATE TABLE orders (id INTEGER PRIMARY KEY, user_id INTEGER, total INTEGER)",
    );
    tx.commit();
  }

  // Success path: order INSERT and job enqueue commit together.
  {
    const tx = db.transaction();
    tx.execute(
      "INSERT INTO orders (user_id, total) VALUES (?, ?)",
      [42, 9900],
    );
    tx.query(
      "SELECT honker_enqueue(?, ?, ?, ?, ?, ?, ?)",
      ['emails', JSON.stringify({ to: 'alice@example.com', order_id: 42 }),
       null, null, 0, 3, null],
    );
    tx.commit();
  }

  {
    const orders = db.query("SELECT id FROM orders");
    const queued = db.query("SELECT payload FROM _honker_live WHERE queue='emails'");
    console.log(`committed: ${orders.length} order(s), ${queued.length} job(s)`);
    assert.equal(orders.length, 1);
    assert.equal(queued.length, 1);
  }

  // Rollback path: business logic fails, both writes disappear.
  try {
    const tx = db.transaction();
    tx.execute(
      "INSERT INTO orders (user_id, total) VALUES (?, ?)",
      [43, 5000],
    );
    tx.query(
      "SELECT honker_enqueue(?, ?, ?, ?, ?, ?, ?)",
      ['emails', JSON.stringify({ to: 'bob@example.com', order_id: 43 }),
       null, null, 0, 3, null],
    );
    throw new Error("boom — simulated payment-processing failure");
    // tx.commit() would have been here, but the throw jumps to catch
  } catch (e) {
    console.log(`rolled back: ${e.message}`);
    // Transaction auto-rolls back when the object goes out of scope
    // without .commit(). We don't need to do anything here.
  }

  {
    const orders = db.query("SELECT id FROM orders");
    const queued = db.query("SELECT payload FROM _honker_live WHERE queue='emails'");
    console.log(`after rollback: ${orders.length} order(s), ${queued.length} job(s)`);
    assert.equal(orders.length, 1);
    assert.equal(queued.length, 1);
  }

  console.log("atomic enqueue + rollback both work as expected");
} finally {
  fs.rmSync(dir, { recursive: true, force: true });
}

File: packages/honker/examples/atomic.py
"""Atomic business-write + enqueue in one transaction.

This is honker's killer feature vs Redis/Celery: the job enqueue
commits in the same SQLite transaction as your business INSERT.
Rollback drops both. No dual-write window, no outbox pattern needed.

    python examples/atomic.py
"""

import os
import tempfile

import honker


def main():
    with tempfile.TemporaryDirectory() as d:
        db = honker.open(os.path.join(d, "app.db"))

        # Set up a plain business table. Nothing honker-specific.
        with db.transaction() as tx:
            tx.execute(
                "CREATE TABLE orders (id INTEGER PRIMARY KEY, user_id INTEGER, total INTEGER)"
            )

        emails = db.queue("emails")

        # Success path: order INSERT and job enqueue commit together.
        with db.transaction() as tx:
            tx.execute(
                "INSERT INTO orders (user_id, total) VALUES (?, ?)",
                [42, 9900],
            )
            emails.enqueue({"to": "alice@example.com", "order_id": 42}, tx=tx)

        orders = db.query("SELECT id, user_id, total FROM orders")
        queued = db.query("SELECT payload FROM _honker_live WHERE queue='emails'")
        print(f"committed: {len(orders)} order(s), {len(queued)} job(s)")
        assert len(orders) == 1 and len(queued) == 1

        # Rollback path: raise inside the block, both writes disappear.
        try:
            with db.transaction() as tx:
                tx.execute(
                    "INSERT INTO orders (user_id, total) VALUES (?, ?)",
                    [43, 5000],
                )
                emails.enqueue({"to": "bob@example.com", "order_id": 43}, tx=tx)
                raise RuntimeError("boom — simulate a payment-processing failure")
        except RuntimeError as e:
            print(f"rolled back: {e}")

        orders = db.query("SELECT id FROM orders")
        queued = db.query("SELECT payload FROM _honker_live WHERE queue='emails'")
        print(f"after rollback: {len(orders)} order(s), {len(queued)} job(s)")
        # Still just the first order + first job; the rolled-back pair is gone.
        assert len(orders) == 1 and len(queued) == 1

        print("atomic enqueue + rollback both work as expected")


if __name__ == "__main__":
    main()
